# WT Vector Cloning 

## Homology Arm Insert

### PCR amplify HA region (twist region + homology arms)

| Reagent | Volume |
| :--- | :--- |
| WT Hap1 gDNA	| 100ng |
| HA Primer F (10uM)	| 1.5ul |
| HA Primer R (10uM)	| 1.5ul |
| Kapa Hifi MM |	25ul |
| H2O	| to 50ul |

1.	Run PCR protocol under Rebecca’s name 
2.	Run on gel
3.	Excise bands 
4.	Gel purify (elute in 12ul EB buffer) 
5.	Nanodrop 

### Digest with NotI and SbfI:

| Reagent | Volume |
| :--- | :--- |
| Purified insert |	11ul |
| SbfI	| 1ul |
|NotI |	1ul |
|Cutsmart Buffer |	2ul |
|H2O |	to 20ul |

1.	Keep at 37C for at least 1hr, 80C for 20min



## WT Backbone:

| Reagent | Volume |
| :--- | :--- |
| Uncut pmin plasmid |	4000ng |
| SbfI	| 1ul |
| NotI |	1ul |
| Cutsmart Buffer |	2ul |
| H2O |	to 20ul |

1.	Keep at 37C for at least 1hr, 80C for 20min
2.	Run on a 1% gel 
3.	Excise band 
4.	Gel purify according to own protocol 
5.	Nanodrop






## Backbone and insert ligation: 

•	Calculate insert concentrations needed for 3:1 and 1:1 ratios of insert to vector on NEBiocalculator 
o	Assumes 1ul vector 

Set up:
DNA (insert)	*NEB calculator output 
Purified / digested Backbone 	1ul
Ligase 	1ul
Ligase buffer 	2ul 
H2O	To 20ul

1.	Incubate at room temp for 10min 
2.	Heat inactivate for 10min at 65C
3.	Place on ice 


Transformation: 

1.	Thaw NEB-beta E. coli cells on ice
2.	Create tubes containing 50ul cells
3.	Add 2ul of ligation mix to tubes
o	Flick to mix
4.	Place on ice for 30min
o	Preheat block to 42C
5.	Incubate at 42C for 30sec exactly
6.	Place on ice for 5min
o	Set block to 37C
7.	Add 950ul of SOM to samples
8.	Shake for 1hr at 37C at 650rpm
9.	Create 2 plates per sample: 30ul and 300ul 
10.	Incubate at 37C overnight 

Select colonies 
•	Incubate overnight 

Miniprep 

Notes:
•	Prior to sequencing, perform a diagnostic digest on 2ul of samples 
DNA	2ul
NspI	1ul
Cutsmart buffer 	2ul
H2O	To 20ul
•	37C for 1hr
•	NspI selected based off band patterns on simulated Snap gene gel 
•	Looked for RE that would produce identifiable band patterns for all samples 

Sequence 
•	Sanger sequence to validate constructs 
•	Use minBB and HDRWT primers F/R (for tubes per sample)
•	Check traces carefully to check for introduced variants (accounting for Hap1 variants)




